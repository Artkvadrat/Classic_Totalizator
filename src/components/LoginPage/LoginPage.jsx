import React, { useMemo, useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { Form, Input, Button, Row } from 'antd';

import {
  loginUser,
  testJwtTokenFromLocalStorage
} from '../../ducks/loginPage/loginPage';

const LoginPage = () => {
  const dispatch = useDispatch();

  useEffect(() => {
    const token = window.localStorage.getItem('jwtToken');
    if (token) {
      dispatch(testJwtTokenFromLocalStorage(token));
    }
  }, [dispatch]);

  const submitLoginForm = useMemo(
    () => (values) => {
      const data = {
        email: values.email,
        password: values.password
      };

      dispatch(loginUser(data));
    },
    [dispatch]
  );

  return (
    <Row
      type="flex"
      justify="center"
      align="middle"
      style={{ textAlign: 'center', paddingTop: '200px' }}
    >
      <Form layout="vertical" onFinish={submitLoginForm}>
        <Form.Item
          label="Email"
          name="email"
          rules={[
            {
              required: true,
              message: 'Please enter email'
            }
          ]}
        >
          <Input placeholder="Email" />
        </Form.Item>
        <Form.Item
          label="Password"
          name="password"
          rules={[
            {
              required: true,
              message: 'Please enter password'
            }
          ]}
        >
          <Input.Password placeholder="Password" />
        </Form.Item>
        <Form.Item>
          <Button type="primary" htmlType="submit">
            Submit
          </Button>
        </Form.Item>
      </Form>
    </Row>
  );
};

export default LoginPage;
