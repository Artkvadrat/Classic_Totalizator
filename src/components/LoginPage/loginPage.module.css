.backButton {
    margin: 50px;
    background-color: inherit;
    border: 0;
    font-size: 20px;
    cursor: pointer;
}

.backButton:focus {
    outline: none;
}

.loginForm {
    display: flex;
    flex-direction: column;
    width: 500px;
    margin: 200px auto;
}

.loginForm input {
    padding: 10px;
    margin-bottom: 20px;
}

.loginForm button {
    margin: 0 auto;
    width: 100px;
    background-color: #f3f30a;
    border: 0;
    padding: 10px 15px;
    border-radius: 10px;
}