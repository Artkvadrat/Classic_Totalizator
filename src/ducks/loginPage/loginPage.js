import HTTPService from '../../services/HTTPService/HTTPService';

export const REQUESTED_JWT_TOKEN = 'login_page/requested';
export const RECEIVED_JWT_TOKEN = 'login_page/received';
export const TESTED_JWT_TOKEN = 'login_page/tested';
export const DENIED_JWT_TOKEN = 'login_page/denied';
export const LOGOUT = 'login_page/logout';

const requestedJwtToken = () => ({
  type: REQUESTED_JWT_TOKEN
});

const receivedJwtToken = (data) => ({
  type: RECEIVED_JWT_TOKEN,
  payload: data
});

const testedJwtToken = (data) => ({
  type: TESTED_JWT_TOKEN,
  payload: data
});

const deniedJwtToken = () => ({
  type: DENIED_JWT_TOKEN
});

const logoutAction = () => ({
  type: LOGOUT
});

export const loginUser = ({ email, password }) => (dispatch) => {
  dispatch(requestedJwtToken());

  return HTTPService.request({
    method: 'POST',
    path: '/api/Auth/admin/login',
    body: {
      login: email,
      password
    }
  })
    .then(({ jwtString }) => {
      dispatch(receivedJwtToken(jwtString));
    })
    .catch(() => {
      dispatch(deniedJwtToken());
    });
};

export const testJwtTokenFromLocalStorage = (jwtToken) => (dispatch) => {
  dispatch(requestedJwtToken());

  return HTTPService.request({ path: '/api/Events/outcomes' })
    .then(() => {
      dispatch(testedJwtToken(jwtToken));
    })
    .catch(() => {
      dispatch(deniedJwtToken());
    });
};

export const logout = () => (dispatch) => {
  dispatch(logoutAction());
};

const initialState = {
  isLoading: false,
  isLoggedIn: false
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case REQUESTED_JWT_TOKEN:
      return {
        ...state,
        isLoading: true
      };
    case RECEIVED_JWT_TOKEN:
      window.localStorage.setItem('jwtToken', action.payload);
      return {
        ...state,
        isLoading: false,
        isLoggedIn: true
      };
    case TESTED_JWT_TOKEN:
      return {
        ...state,
        isLoading: false,
        isLoggedIn: true
      };
    case DENIED_JWT_TOKEN:
      window.localStorage.clear();
      return {
        ...state,
        isLoading: false,
        isLoggedIn: false
      };
    case LOGOUT:
      window.localStorage.clear();
      return {
        ...state,
        isLoggedIn: false
      };
    default:
      return state;
  }
};

export default reducer;
