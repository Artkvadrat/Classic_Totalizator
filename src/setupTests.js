import enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-17-updated';

enzyme.configure({ adapter: new Adapter() });

const localStorageMock = {
  getItem: jest.fn(),
  setItem: jest.fn(),
  removeItem: jest.fn(),
  clear: jest.fn()
};
window.localStorage = localStorageMock;
