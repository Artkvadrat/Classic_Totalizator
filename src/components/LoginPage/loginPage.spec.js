import React from 'react';
import { shallow } from 'enzyme';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory } from 'react-router-dom';
import { loginUser } from '../../ducks/loginPage/loginPage';

import '../../setupTests';

import LoginPage from './LoginPage';

jest.mock('react-router-dom', () => ({
  useHistory: jest.fn()
}));

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: () => ({
    isLoggedIn: false
  })
}));

jest.mock('../../ducks/loginPage/loginPage', () => ({
  loginUser: jest.fn()
}));

describe('LoginPage component', () => {
  let wrapper;
  const history = {
    push: jest.fn()
  };

  const dispatch = jest.fn();

  const mockedLoginUser = jest.fn();

  beforeEach(() => {
    useHistory.mockReturnValue(history);
    useDispatch.mockReturnValue(dispatch);
    loginUser.mockReturnValue(mockedLoginUser);

    wrapper = shallow(<LoginPage />);
  });

  it('should render without error', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render button without error', () => {
    expect(wrapper.find('button.backButton')).toBeDefined();
    expect(wrapper.find('button.backButton')).toMatchSnapshot();
  });

  it('should render form without error', () => {
    expect(wrapper.find('form.loginForm')).toBeDefined();
    expect(wrapper.find('form.loginForm')).toMatchSnapshot();
  });

  it('should trigger history.push on click button', () => {
    const button = wrapper.find('button.backButton');
    button.simulate('click');

    expect(history.push).toHaveBeenCalled();
    expect(history.push).toHaveBeenCalledWith('/');
  });

  it('should trigger dispatch on submit form', () => {
    const fakeEvent = {
      preventDefault: jest.fn(),
      target: [{ value: 'email' }, { value: 'password' }]
    };

    const form = wrapper.find('form.loginForm');
    form.simulate('submit', fakeEvent);

    const expectedData = {
      email: 'email',
      password: 'password'
    };

    expect(fakeEvent.preventDefault).toHaveBeenCalled();
    expect(fakeEvent.target[0]).toEqual({ value: 'email' });
    expect(fakeEvent.target[1]).toEqual({ value: 'password' });
    expect(dispatch).toHaveBeenCalled();
    expect(dispatch).toHaveBeenCalledWith(loginUser());
    expect(loginUser).toHaveBeenCalled();
    expect(loginUser).toHaveBeenCalledWith(expectedData);
  });
});
