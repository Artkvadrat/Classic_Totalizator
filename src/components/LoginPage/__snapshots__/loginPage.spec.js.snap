// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage component should render button without error 1`] = `
<button
  className="backButton"
  onClick={[Function]}
  type="button"
>
  Назад
</button>
`;

exports[`LoginPage component should render form without error 1`] = `
<form
  className="loginForm"
  onSubmit={[Function]}
>
  <p>
    Email
  </p>
  <input
    type="text"
  />
  <p>
    Password
  </p>
  <input
    type="password"
  />
  <button
    type="submit"
  >
    Login
  </button>
</form>
`;

exports[`LoginPage component should render without error 1`] = `
<div>
  <button
    className="backButton"
    onClick={[Function]}
    type="button"
  >
    Назад
  </button>
  <form
    className="loginForm"
    onSubmit={[Function]}
  >
    <p>
      Email
    </p>
    <input
      type="text"
    />
    <p>
      Password
    </p>
    <input
      type="password"
    />
    <button
      type="submit"
    >
      Login
    </button>
  </form>
</div>
`;
